flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sources/TableSourceUtil.scala

```diff
@@ -173,16 +174,30 @@ object TableSourceUtil {
         fieldTypes = fieldTypes.patch(idx, Seq(proctimeType), 1)
       }
     }
-    val (selectedFieldNames, selectedFieldTypes) =
+    var (selectedFieldNames, selectedFieldTypes) =
       if (selectedFields.isDefined) {
         // filter field names and types by selected fields
-        (
-          selectedFields.get.map(fieldNames(_)),
-          selectedFields.get.map(fieldTypes(_))
-        )
+        (selectedFields.get.map(fieldNames(_)), selectedFields.get.map(fieldTypes(_)))
       } else {
         (fieldNames, fieldTypes)
       }
+
+    // patch rowtime field according to WatermarkSpec
+    if (streaming && watermarkSpec.isDefined) {
+      // TODO: [FLINK-14473] we only support top-level rowtime attribute right now
```

Comment:
Throw exception early if we do not support it in the patch.

Reply:
This is the earliest place to throw this exception, otherwise, we should throw exception in `TableSchema`.